docs/advanced/developer-guide.md

```diff
@@ -0,0 +1,303 @@
+
+### Customizing the Build
+
+There are several Makefile variables that control the build process and the
+name of the resulting container image. The following are targeted targeted for
+build customization and they can be specified via environment variables or
+makefile overrides.
+
+| Variable                   | Description                                                       | Default value
+| -------------------------- | ----------------------------------------------------------------- | ----------- |
+| HOSTMOUNT_PREFIX           | Prefix of system directories for feature discovery (local builds) | / (*local builds*) /host- (*container builds*)
+| IMAGE_BUILD_CMD            | Command to build the image                                        | docker build
+| IMAGE_BUILD_EXTRA_OPTS     | Extra options to pass to build command                            | *empty*
+| IMAGE_PUSH_CMD             | Command to push the image to remote registry                      | docker push
+| IMAGE_REGISTRY             | Container image registry to use                                   | k8s.gcr.io/nfd
+| IMAGE_TAG_NAME             | Container image tag name                                          | &lt;nfd version&gt;
+| IMAGE_EXTRA_TAG_NAMES      | Additional container image tag(s) to create when building image   | *empty*
+| K8S_NAMESPACE              | nfd-master and nfd-worker namespace                               | kube-system
+| KUBECONFIG                 | Kubeconfig for running e2e-tests                                  | *empty*
+| E2E_TEST_CONFIG            | Parameterization file of e2e-tests (see [example](test/e2e/e2e-test-config.exapmle.yaml)) | *empty*
+
+For example, to use a custom registry:
+
+```bash
+make IMAGE_REGISTRY=<my custom registry uri>
+```
+
+Or to specify a build tool different from Docker, It can be done in 2 ways:
+
+1. via environment
+```bash
+IMAGE_BUILD_CMD="buildah bud" make
+```
+1. by overriding the variable value
+```bash
+make  IMAGE_BUILD_CMD="buildah bud"
+```
+
+### Testing
+
+Unit tests are automatically run as part of the container image build. You can
+also run them manually in the source code tree by simply running:
+
+```bash
+make test
+```
+
+End-to-end tests are built on top of the e2e test framework of Kubernetes, and,
+they required a cluster to run them on. For running the tests on your test
+cluster you need to specify the kubeconfig to be used:
+
+```bash
+make e2e-test KUBECONFIG=$HOME/.kube/config
+```
+
+## Running Locally
+
+You can run NFD locally, either directly on your host OS or in containers for
+testing and development purposes. This may be useful e.g. for checking
+features-detection.
+
+### NFD-Master
+
+When running as a standalone container labeling is expected to fail because
+Kubernetes API is not available. Thus, it is recommended to use `--no-publish`
+command line flag. E.g.
+
+```bash
+$ docker run --rm --name=nfd-test <NFD_CONTAINER_IMAGE> nfd-master --no-publish
```

Comment:
The same comment as for the nfd-worker, can we use here ${NFD_CONTAINER_IMAGE} for easier copy and waste?

Reply:
Changed